src/main/java/org/apache/maven/plugins/pmd/exec/PmdErrorMessageReporter.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.plugins.pmd.exec;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import net.sourceforge.pmd.util.log.PmdReporter;
+import org.apache.maven.reporting.MavenReportException;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.event.Level;
+
+class PmdErrorMessageReporter implements PmdReporter {
+    private final PmdReporter delegate;
+    private List<Throwable> errors = new ArrayList<>();
+
+    PmdErrorMessageReporter(PmdReporter delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public boolean isLoggable(Level level) {
+        return delegate.isLoggable(level);
+    }
+
+    @Override
+    public void logEx(Level level, @Nullable String message, Object[] formatArgs, @Nullable Throwable error) {
+        if (error != null) {
+            if (error instanceof RuntimeException && error.getCause() instanceof MavenReportException) {
+                errors.add(error.getCause());
+            } else {
+                errors.add(error);
+            }
+        }
+
+        delegate.logEx(level, message, formatArgs, error);
+    }
+
+    @Override
+    public int numErrors() {
+        return delegate.numErrors();
+    }
+
+    public void rethrowErrorsAsMavenReportException() throws MavenReportException {
+        if (errors.isEmpty()) {
+            return;
+        }
+
+        final MavenReportException exception;
+
+        Throwable firstError = errors.get(0);
```

Comment:
I am bit confused, you log multiple issue -- yet this one processes only the first one? Can multiple ones happen? What ist the purpose to log them if they are rethrown and printed by the system anyway? Doesn't this produce duplicate output?

@elharo

Reply:
@elharo What is your PoV on this approach?

Reply:
I'm not sure I see the confusion here. If I follow, multiple exceptions are thrown and stored while the process runs. When it's done a single exception is thrown, which is either the first exception or a wrapper around the first exception. The other exceptions are stored ion addSuppressed below, and should be visible in the eventual stack trace. Is that correct?

Reply:
OK, I don't immediately see how that happens here, but if it does I would prefer not to do that. In general, log or throw, not both.

Reply:
Look at logEx. There is your log.

Reply:
Multiple exception can potentially occur. E.g. if someone tries to run CPD on java files, that don't compile and are not valid java source - each file would produce an exception.

And right, as I delegate it to the original PmdReporter, it will have been logged (through slf4j) already, when we rethrow the exception(s).

The whole goal of this exercise is, that the build fails, when some exceptions occurred, that potentially mean, that not all files have been analyzed. Who reads the logs, if in the end it says "BUILD SUCCESS"?

When this plugin was created, it never failed when there was any problem while executing PMD - some files where then just silently not analyzed. For the pmd:pmd goal, the parameter "skipPmdError" was introduced with the sake for compatibility defaulting to the old behavior (just ignoring any errors). The cpd:cpd goal doesn't have such an option and I don't know how this behaved earlier. But if any errors occur during analysis, IMHO the build must fail.

Reply:
I've now done another approach: just let PMD log the errors and m-pmd-p fails the build with the message "There were ... errors while executing CPD". The actual errors (e.g. "Lexical error in file ...") have been already logged, so users just need to scroll up. This also means, that these errors are logged only once and we don't need a custom PmdErrorMessageReporter.

> @adangel Do you think you could add a test which invokes the exception handling and we can see the log output and `throw` combined?

I've added a test case, see CpdReportTest#testWithCpdErrors. It asserts that we get the main exception which fails the build as well as the actual errors appear in the log output.